packages/node/src/indexer/unfinalizedBlocks.service.ts

```diff
@@ -0,0 +1,289 @@
+      if (lastVerifiableBlock.hash !== this.finalizedHeader.hash) {
+        logger.warn(
+          `Block fork found, enqueued un-finalized block at ${lastVerifiableBlock.blockHeight} with hash ${lastVerifiableBlock.hash}, actual hash is ${this.finalizedHeader.hash}`,
+        );
+        return this.finalizedHeader;
+      }
+    } else {
+      // Unfinalized blocks below finalized block
+      let header = this.finalizedHeader;
+      /*
+       * Iterate back through parent hashes until we get the header with the matching height
+       * We use headers here rather than getBlockHash because of potential caching issues on the rpc
+       * If we're off by a large number of blocks we can optimise by getting the block hash directly
+       */
+      if (
+        header.number - lastVerifiableBlock.blockHeight >
+        UNFINALIZED_THRESHOLD
+      ) {
+        header = await this.api.getBlockByHeightOrHash(
+          lastVerifiableBlock.blockHeight,
+        );
+      } else {
+        while (lastVerifiableBlock.blockHeight !== header.number) {
+          header = await this.api.getBlockByHeightOrHash(header.parentHash);
+        }
+      }
+
+      if (header.hash !== lastVerifiableBlock.hash) {
+        logger.warn(
+          `Block fork found, enqueued un-finalized block at ${lastVerifiableBlock.blockHeight} with hash ${lastVerifiableBlock.hash}, actual hash is ${header.hash}`,
+        );
+        return header;
+      }
+    }
+
+    return;
+  }
+
+  private async saveUnfinalizedBlocks(
+    unfinalizedBlocks: UnfinalizedBlocks,
+    tx: Transaction,
+  ): Promise<void> {
+    return this.setMetadata(
+      METADATA_UNFINALIZED_BLOCKS_KEY,
+      JSON.stringify(unfinalizedBlocks),
+      tx,
+    );
+  }
+
+  private async getLastCorrectFinalizedBlock(
+    forkedHeader: Block,
+  ): Promise<number | undefined> {
+    const bestVerifiableBlocks = this.unfinalizedBlocks.filter(
+      ([bestBlockHeight]) =>
+        Number(bestBlockHeight) <= this.finalizedBlockNumber,
+    );
+
+    let checkingHeader = forkedHeader;
+
+    // Work backwards through the blocks until we find a matching hash
+    for (const [block, hash] of bestVerifiableBlocks.reverse()) {
+      if (hash === checkingHeader.hash || hash === checkingHeader.parentHash) {
+        return Number(block);
+      }
+
+      // Get the new parent
+      checkingHeader = await this.api.getBlockByHeightOrHash(
+        checkingHeader.parentHash,
+      );
+    }
+
+    return this.lastCheckedBlockHeight;
+  }
+
+  async resetUnfinalizedBlocks(tx: Transaction): Promise<void> {
+    await this.setMetadata(METADATA_UNFINALIZED_BLOCKS_KEY, '[]', tx);
+    this.unfinalizedBlocks = [];
+  }
+
+  async resetLastFinalizedVerifiedHeight(tx: Transaction): Promise<void> {
+    return this.setMetadata(METADATA_LAST_FINALIZED_PROCESSED_KEY, null, tx);
+  }
+
+  private async setMetadata(
+    key: string,
+    value: string | number | boolean,
```

Comment:
this might need to align with current metadata table changes
https://github.com/subquery/subql/blob/932f6a0c6358743b48c671a60faa8e608d79fd15/packages/node/src/indexer/unfinalizedBlocks.service.ts#L291